src/cargo/util/lints.rs

```diff
@@ -880,6 +882,111 @@ pub fn unused_dependencies(
     Ok(())
 }
 
+pub fn unexpected_target_cfgs(
+    ws: &Workspace<'_>,
+    pkg: &Package,
+    path: &Path,
+    error_count: &mut usize,
+    gctx: &GlobalContext,
+) -> CargoResult<()> {
+    fn warn_on_unexpected_cfgs(
+        gctx: &GlobalContext,
+        check_cfg: &CheckCfg,
+        cfg_expr: &CfgExpr,
+        lint_level: LintLevel,
+        error_count: &mut usize,
+        path: Option<&Path>,
+        suffix: &str,
+    ) -> CargoResult<()> {
+        let prefix = if let Some(path) = path {
+            format!("{path}: ", path = path.display())
+        } else {
+            "".to_string()
+        };
+
+        let mut emit_lint = |msg| match lint_level {
+            LintLevel::Warn => gctx.shell().warn(msg),
+            LintLevel::Deny | LintLevel::Forbid => {
+                *error_count += 1;
+                gctx.shell().error(msg)
+            }
+            LintLevel::Allow => Ok(()),
+        };
+
+        cfg_expr.try_for_each(|cfg| -> CargoResult<()> {
```

Comment:
Why a `for_each` rather than just a `iter` and using a `for` loop?

Reply:
I think that would require having generators (nightly-only) or a very heavy manual state machine in a `CfgExpr`-iterator.

Which I don't is something we want for maintainability.

Reply:
I assume the answer then is "its walking a recursive data structure and its easier to write it that way".  btw I think then the name `walk` would better clarify intent.